TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -472,6 +486,91 @@ - (NSArray *)parseDictionaryOfSchemasForKey:(NSString *)key schema:(NSDictionary
 
 #pragma mark - Keyword-specific parser methods
 
+// Valid values for type are:
+//    A. Nil (type keyword not present)
+//    B. An array of strings that are all valid JSON types
+//    C. A string that is a valid JSON type
+// Returns an array of strings; indirectly returns boolean indicating whether type was explicit or implied by keywords
+- (NSArray *)parseTypeKeywordForSchema:(NSDictionary *)schema explicit:(BOOL *)isTypeSpecified
+{
+    [self pushPathComponent:TWTJSONSchemaKeywordType];
+    id type = schema[TWTJSONSchemaKeywordType];
+
+    if (isTypeSpecified) {
+        *isTypeSpecified = type != nil;
+    }
+
+    // Case A: Type keyword not present
+    if (!type) {
+        // Check if type-specific keywords exist
+        return [self impliedTypeForSchema:schema];
+    }
+
+    [self failIfObject:type isNotKindOfOneOfClasses:[NSString class], [NSArray class], nil];
+
+    if ([type isKindOfClass:[NSArray class]]) {
+
+        // Case B: Type is an array
+        [(NSArray *)type enumerateObjectsUsingBlock:^(NSString *typeString, NSUInteger index, BOOL *stop) {
+            [self pushPathComponent:@(index)];
+            [self failIfObject:typeString isNotMemberOfSet:[self validJSONTypeKeywords]];
+            [self popPathComponent];
+        }];
+    } else {
+        // Case C: Type is a string
+        [self failIfObject:type isNotMemberOfSet:[self validJSONTypeKeywords]];
+        type = @[ type ];
+    }
+
+    [self popPathComponent];
+    
+    return type;
+}
+
+
+// If type keyword is not present, determines whether type-specific keywords indicate one or more types
+- (NSArray *)impliedTypeForSchema:(NSDictionary *)schema
+{
+    static NSSet *objectKeywords = nil;
+    static NSSet *arrayKeywords = nil;
+    static NSSet *stringKeywords = nil;
+    static NSSet *numberKeywords = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        objectKeywords = [NSSet setWithObjects:TWTJSONSchemaKeywordMaxProperties, TWTJSONSchemaKeywordMinProperties, TWTJSONSchemaKeywordRequired,
+                          TWTJSONSchemaKeywordProperties, TWTJSONSchemaKeywordAdditionalProperties, TWTJSONSchemaKeywordPatternProperties,
+                          TWTJSONSchemaKeywordDependencies, nil];
```

Comment:
Indent these lines to line up with `TWTJSONSchemaKeywordMaxProperties`. Make sure we don’t exceed 160 characters. Same on the lines below.